Projectfile

amjith opened the pull request:
## Description
This will make the `init` command and the `new` command interactive.


## Checklist
<!--- We appreciate your help and please add yourself to the AUTHORS file. -->

amjith commented:
I'll fix the conflicts and make the tests pass.

ofek commented:
Thanks! I'll look this over later after I fix master. My initial impression though is I would prefer to enable interactivity with an `-i/--interactive` flag. Only being interactive makes the config file way less useful and as a user I would want non-interactive by default.

amjith commented:
I think you make a good point. Except the package name and the description all the other questions are pretty static and it can be supplied by the config file. 

I'll change the behavior to add a `-i` option for interactive.

amjith commented:
The tests are now passing. But I haven't added new tests to cover the code that I have added. I'll work on adding tests tomorrow. In the meantime, I welcome any comments or suggestions.

amjith commented:
I've added tests to the interactive init command. I have also fixed the failing tests. The PR is pretty big. I'm happy to do a screen share and walk you through the main points if you're interested. 

I'll add more tests to the interactive new command as well, but I think this PR is ready for your review.

ofek commented on hatch/settings.py:
```diff
@@ -19,8 +20,8 @@
     ('pypi_username', ''),
-    ('name', 'U.N. Owen'),
-    ('email', 'me@un.known'),
+    ('author', get_user() or 'U.N. Owen'),
```
I understand the decision here, but can we please keep it as before? It would be a breaking change.

ofek commented:
In init and new command files can you please change `from hatch.settings import load_settings` to `from hatch.settings import copy_default_settings, load_settings` and use that new function in the `except FileNotFoundError:`? That was an oversight of mine.

You can then use the default settings in the functions.

ofek commented:
After those changes, we also need to make ourselves the `tool.hatch` as we discussed.

Here are some examples I've found:

https://github.com/pypa/pip/blob/54b983c2bd56dfbd61bbcc2dcc9177d4e55e25af/pyproject.toml
https://github.com/saezlab/pypath/blob/b285a3a80815a6f27100966c0ad5da01fe33fe61/pyproject.toml
https://github.com/pandas-dev/pandas/blob/e0fe5cc60b1dc0d777223bba64b8abfc0e0e02ab/pyproject.toml
https://github.com/twisted/twisted/blob/2dde252439267573fb802b5ab64af9517b3ea856/pyproject.toml

ofek commented:
https://github.com/pypa/pip/pull/4144#issuecomment-332024594

ofek commented on tests/files/licenses/test_apache2.py:
```diff
@@ -25,7 +25,7 @@ def test_name():
         settings = copy_default_settings()
         settings['licenses'] = ['apache2']
-        settings['name'] = 'Don Quixote'
+        settings['author'] = 'Don Quixote'
```
here too

ofek commented on tests/files/licenses/test_cc0.py:
```diff
@@ -25,7 +25,7 @@ def test_name():
         settings = copy_default_settings()
         settings['licenses'] = ['cc0']
-        settings['name'] = 'Guy Fawkes'
+        settings['author'] = 'Guy Fawkes'
```
here too

ofek commented on tests/files/licenses/test_mit.py:
```diff
@@ -25,7 +25,7 @@ def test_name():
         settings = copy_default_settings()
         settings['licenses'] = ['mit']
-        settings['name'] = 'Don Quixote'
+        settings['author'] = 'Don Quixote'
```
here too

ofek commented on tests/files/licenses/test_mpl.py:
```diff
@@ -25,7 +25,7 @@ def test_name():
         settings = copy_default_settings()
         settings['licenses'] = ['mpl']
-        settings['name'] = 'Red Panda'
+        settings['author'] = 'Red Panda'
```
here too

ofek commented:
Sorry, I missed a few.

ofek commented:
Looks great! We just need to change the `pyproject.toml`: https://github.com/pypa/pip/pull/4144#issuecomment-332024594 until end of thread

So, initially I was thinking we could have metadata and commands, but exclude packages for when Hatch supports Pipfiles. Sound good?

amjith commented:
Looks like one test is failing on one Python36-x64 Appveyor. 

Looks to me it's a timeout failure. Can you take a look?

ofek commented:
That's just virtualenv being slow, it happens sometimes :/

ofek commented on hatch/commands/new.py:
```diff
@@ -93,18 +95,30 @@ def new(name, no_env, pyname, pypath, global_packages, env_name, basic, cli, lic
         sys.exit(1)
 
+    if licenses:
+        settings['licenses'] = map(str.strip, licenses.split(','))
```
In `init` this comes after the interactive branch. What do you want to give precedence?

amjith replied:
If the user runs this: 

```
hatch init -i -l 'mit,mpl'
```

Hatch will use the `mit,mpl` as the default value during the interactive mode. If the user chooses to override it during the interactive mode then the overridden value from the interactive mode will be used.

ofek replied:
Right now the thing is that in `new` the `if licenses` comes before the `if interactive` but in `init` it's the opposite.

amjith replied:
Good catch. They were functionally equivalent. But I've changed the code to match the same pattern in both files.

ofek commented on tests/commands/test_init.py:
```diff
@@ -27,6 +28,33 @@ def test_config_not_exist():
+        runner.invoke(hatch, ['init', '-i', '--basic', '-ne'], input='ok\n0.1.0\nTest Description\nPicard\npicard@startrek.com\nmpl\n')
+
+        assert os.path.exists(os.path.join(d, 'ok', '__init__.py'))
+        assert "__version__ = '0.1.0'\n" == open(os.path.join(d, 'ok', '__init__.py')).read()
```
use https://github.com/ofek/hatch/blob/ae3e9d4c4a569955f3fc98d4a07c6f0ffea5e1dc/tests/utils.py#L37

ofek commented on tests/commands/test_new.py:
```diff
@@ -28,6 +29,36 @@ def test_config_not_exist():
+        runner.invoke(hatch, ['new', '-i', '--basic', '-ne'], input='ok\n0.1.0\nTest Description\nPicard\npicard@startrek.com\nmpl\n')
+
+        assert os.path.exists(os.path.join(d, 'ok', 'ok', '__init__.py'))
+        assert "__version__ = '0.1.0'\n" == open(os.path.join(d, 'ok', 'ok', '__init__.py')).read()
```
use https://github.com/ofek/hatch/blob/ae3e9d4c4a569955f3fc98d4a07c6f0ffea5e1dc/tests/utils.py#L37

ofek commented:
The PR looks great except the `pyproject.toml` needs:

```toml
[build-system]
requires = ["setuptools", "wheel"]
```

and put metadata right under `[tool.hatch]`.

ofek commented:
Tests failing due to source -> metadata change.

amjith commented:
WooHoo!! 🎉
